### ISSUE TITLE
USER, HOME Environment Variables not set in Session started by Proxy

### ISSUE THREAD

**Issue by ToppDev** [Xpra-org/xpra#3021]
Hi,

I am running a proxy server under the user `linuxuser`
```
xpra proxy :100 --bind-tcp=0.0.0.0:10000,auth=sqlite:filename=./xpra-auth.sdb --socket-dir=/tmp --html=on
```

To connect I use the html browser and my sqlite file looks like this
```
$ python3 /usr/lib/python3/dist-packages/xpra/server/auth/sqlite_auth.py ./xpra-auth.sdb list
2 rows found:
-----------------------------------------------------------------------------------
| username| password|       uid|       gid| displays| env options| session options|
-----------------------------------------------------------------------------------
|     test|       pw| linuxuser| linuxuser|         |            |                |
|   test60|       pw| linuxuser| linuxuser|       60|            |                |
```
The user `test` automatically spawns a new session whereas the user `test60` doesn't spawn a session, but can connect to a session started via
```
xpra start :60
```

Running the `env` command on both sessions gives me
<details><summary>Session started by Proxy</summary>
<p>

```
MWWM=allwm
SHELL=/bin/bash
WINDOWID=6291469
XPRA_PROXY_START_UUID=27e9cfb94f0d48b6b9140e8daa31abfd
GDK_DPI_SCALE=1
GTK_IM_MODULE=xim
XTERM_VERSION=XTerm(353)
QT4_IM_MODULE=xim
CKCON_X11_DISPLAY=:3
XMODIFIERS=@im=none
XTERM_SHELL=/bin/bash
NO_AT_BRIDGE=1
XPRA_PULSE_SINK_DEVICE_NAME=Xpra-Microphone
PWD=/home/linuxuser
LOGNAME=linuxuser
XDG_SESSION_TYPE=x11
MWNO_RIT=true
XAUTHORITY=/home/linuxuser/.Xauthority
LANG=en_US.UTF-8
LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:
XDG_CURRENT_DESKTOP=Xpra
XTERM_LOCALE=en_US.UTF-8
XPRA_SERVER_LOG=/run/user/1001/xpra/:3.log
DISABLE_IMSETTINGS=1
XPRA_PULSE_SOURCE_DEVICE_NAME=Xpra-Speaker
GTK_CSD=0
IMSETTINGS_MODULE=none
CLUTTER_IM_MODULE=xim
QT_QPA_PLATFORM=xcb
LESSCLOSE=/usr/bin/lesspipe %s %s
TERM=xterm
LESSOPEN=| /usr/bin/lesspipe %s
DISPLAY=:3
SHLVL=2
QT_IM_MODULE=xim
MWNOCAPTURE=true
XPRA_LOG_DIR=/run/user/1001/xpra
UBUNTU_MENUPROXY=
DBUS_SESSION_BUS_PID=24921
QT_X11_NO_NATIVE_MENUBAR=1
XDG_RUNTIME_DIR=/run/user/1001
XPRA_PULSE_SERVER=/run/user/1001/xpra/pulse-3/pulse/native
GDK_BACKEND=x11
PATH=/usr/lib/xpra:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
GDK_SCALE=1
XPRA_SERVER_SOCKET=/run/user/1001/xpra/musculus-3
DBUS_SESSION_BUS_ADDRESS=unix:abstract=/tmp/dbus-qzCmOyZbqr,guid=a7833a532ccd031500999e7f602f873a
QT_SCALE_FACTOR=1
PULSE_SERVER=unix:/run/user/1001/xpra/pulse-3/pulse/native
DBUS_SESSION_BUS_WINDOWID=2097153
_=/usr/bin/env
```

</p>
</details>
<details><summary>Session manually started</summary>
<p>

```
MWWM=allwm
SHELL=/bin/bash
WINDOWID=6291469
GDK_DPI_SCALE=1
GTK_IM_MODULE=xim
XTERM_VERSION=XTerm(353)
QT4_IM_MODULE=xim
CKCON_X11_DISPLAY=:60
XMODIFIERS=@im=none
XTERM_SHELL=/bin/bash
NO_AT_BRIDGE=1
XPRA_PULSE_SINK_DEVICE_NAME=Xpra-Microphone
PWD=/home/linuxuser
LOGNAME=linuxuser
XDG_SESSION_TYPE=x11
MWNO_RIT=true
XAUTHORITY=/home/linuxuser/.Xauthority
MOTD_SHOWN=pam
HOME=/home/linuxuser
LANG=en_US.UTF-8
LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:
XDG_CURRENT_DESKTOP=Xpra
OPENBLAS_MAIN_FREE=1
XTERM_LOCALE=en_US.UTF-8
XPRA_SERVER_LOG=/run/user/1001/xpra/:60.log
DISABLE_IMSETTINGS=1
XPRA_PULSE_SOURCE_DEVICE_NAME=Xpra-Speaker
SSH_CONNECTION=xxx.xx.xxx.91 54080 xxx.xx.xxx.169 22
GTK_CSD=0
IMSETTINGS_MODULE=none
CLUTTER_IM_MODULE=xim
QT_QPA_PLATFORM=xcb
LESSCLOSE=/usr/bin/lesspipe %s %s
XDG_SESSION_CLASS=user
GOTOBLAS_MAIN_FREE=1
TERM=xterm
LESSOPEN=| /usr/bin/lesspipe %s
USER=linuxuser
DISPLAY=:60
SHLVL=2
QT_IM_MODULE=xim
MWNOCAPTURE=true
XPRA_LOG_DIR=/run/user/1001/xpra
UBUNTU_MENUPROXY=
DBUS_SESSION_BUS_PID=25349
QT_X11_NO_NATIVE_MENUBAR=1
XDG_RUNTIME_DIR=/run/user/1001
SSH_CLIENT=xxx.xx.xxx.91 54080 22
XPRA_PULSE_SERVER=/run/user/1001/xpra/pulse-60/pulse/native
XDG_DATA_DIRS=/usr/local/share:/usr/share:/var/lib/snapd/desktop
GDK_BACKEND=x11
PATH=/usr/lib/xpra:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
GDK_SCALE=1
XPRA_SERVER_SOCKET=/run/user/1001/xpra/musculus-60
DBUS_SESSION_BUS_ADDRESS=unix:abstract=/tmp/dbus-I4DFspnQ38,guid=204e443d11f205ae448b1203602f8761
SSH_TTY=/dev/pts/0
QT_SCALE_FACTOR=1
PULSE_SERVER=unix:/run/user/1001/xpra/pulse-60/pulse/native
DBUS_SESSION_BUS_WINDOWID=2097153
_=/usr/bin/env
```

</p>
</details>

The problem is, that the proxy server does not set the `USER` and `HOME` environment variable.

I found a similar issue here #2951, but there is no real solution to the problem.

Is this a configuration problem, or a bug?

Xpra version: xpra v4.1-r28309

**Comment by totaam** [comment]
Since your proxy is not running as root, it doesn't change uid (it can't) and it doesn't modify the environment.
Are `$USER` and `$HOME` set when you run the proxy?
I think that's where it should be done. Or am I missing something?

**Comment by ToppDev** [comment]
The proxy runs under the user `linuxuser` and the new server spawned by the proxy also runs under `linuxuser`, which is correct.

It doesn't have to change the uid, but it empties the environment variables `$USER` and `$HOME` which makes programs crash when accessing `~/`

**Comment by totaam** [comment]
Ah, found it - fixed in 074b063356331dc784d3bc672e4cc44929f6005d

This was meant to prevent the root environment variables from being passed to the proxy instance subprocess.
But I guess that this should not be too much of a problem since the subprocess will be launched with `--uid=` and `--gid=`.
(and the new `saved_env` thing is not used for things like `HOME` or `USER`)

Until new builds are available, you can:
* apply the trivial patch
* run your proxy with `XPRA_PROXY_ENV_WHITELIST=LANG,HOSTNAME,PWD,TERM,SHELL,SHLVL,PATH,USER,HOME`

Feel free to re-open if I've missed something.
